neptune/management/internal/utils.py

```diff
@@ -52,3 +54,60 @@ def normalize_project_name(name: str, workspace: Optional[str] = None):
     )
 
     return f"{extracted_workspace_name}/{extracted_project_name}"
+
+
+class ProjectKeyGenerator:
+    _PROJECT_KEY_MIN_LENGTH_FOR_DEFAULT_GENERATOR = 3  # project_name is at least 3 characters long
+    _PROJECT_KEY_MAX_LENGTH_FOR_DEFAULT_GENERATOR = 5
+    _PROJECT_KEY_MIN_NUMBER_ITERATOR = 2
+    _PROJECT_KEY_MAX_NUMBER_ITERATOR = 3
+    _PROJECT_KEY_RANDOM_SUFFIX_SIZE = 3
+    _PROJECT_KEY_RANDOM_SUFFIX_ALPHABET = list(string.ascii_uppercase)
+
+    def __init__(self, project_name: str, existing_project_keys: Set[str]):
+        self.project_name = project_name
+        self.existing_project_keys = existing_project_keys
+        self.max_substring_length = min(
+            len(project_name), self._PROJECT_KEY_MAX_LENGTH_FOR_DEFAULT_GENERATOR
+        )
+
+    def get_default_project_key(self) -> str:
+        unique_key = self._find_unique_project_key("")
+        if unique_key is not None:
+            return unique_key
+
+        #  if above code do not find a proper id then add incrementation
+        unique_key_with_incrementation = self._find_unique_project_key_with_incrementation()
+        if unique_key_with_incrementation is not None:
+            return unique_key_with_incrementation
+
+        #  if above code do not find a proper id then add random string as suffix
+        while 1:
+            project_name_substring = (
+                self.project_name[: self.max_substring_length] + self._generate_random_suffix()
+            )
+            if project_name_substring not in self.existing_project_keys:
+                return project_name_substring
+
+    def _generate_random_suffix(self):
+        alphabet_copy = self._PROJECT_KEY_RANDOM_SUFFIX_ALPHABET.copy()
+        random.shuffle(alphabet_copy)
+        return "".join(alphabet_copy[: self._PROJECT_KEY_RANDOM_SUFFIX_SIZE])
+
+    def _find_unique_project_key_with_incrementation(self) -> Optional[str]:
+        for name_incrementation in range(
+            self._PROJECT_KEY_MIN_NUMBER_ITERATOR, self._PROJECT_KEY_MAX_NUMBER_ITERATOR + 1
+        ):
+            unique_key = self._find_unique_project_key(str(name_incrementation))
```

Comment:
In fact we're sure that we should not check all possible prefixes in `_find_unique_project_key` since we did it before in `get_default_project_key`, but it's really small number of additional iteration on data in memory, so we don't care...

Reply:
clarified on call